File: android/app/src/main/java/com/nhs/online/nhsonline/support/fileDownloading/Base64File.kt
package com.nhs.online.nhsonline.support.fileDownloading

import android.util.Base64
import com.nhs.online.nhsonline.services.knownservices.enums.JavaScriptInteractionMode

class Base64File(
        val fileName: String,
        val fileMimeType: String,
        val encoding: String,
        val source: JavaScriptInteractionMode? = null
) {
    private val DATA_SCHEME = "data:"

    val dataMediaType: String
    val data: String
    val createdFromDataScheme: Boolean
        get() = this.dataMediaType != ""

    init {
        if (encoding.startsWith(DATA_SCHEME, true)) {
            val clobWithoutScheme = encoding.substring(DATA_SCHEME.length)
            // split to place file mime type into first element
            val clobSplitOnComma = clobWithoutScheme.split(',')
            // this should contain ';base64' at end, so strip that as well
            dataMediaType = clobSplitOnComma[0].split(';')[0]
            data = clobSplitOnComma[1]
        }
        else {
            dataMediaType = ""
            data = encoding
        }

    }

    fun decode(): ByteArray {
        return Base64.decode(data, Base64.DEFAULT)
    }
}

File: ios/NHSOnline/NHSOnline/Support/FileDownload/Base64File.swift
import Foundation

enum Base64FileError: Error {
    case failDecoding
}

class Base64File {
    private let DATA_SCHEME = "data:"
    let fileName: String
    let fileMimeType: String
    let dataMediaType: String
    let data: String
    
    public init( fileName: String,
                 mimeType: String,
                 encoding: String ) {
        self.fileName = fileName
        self.fileMimeType = mimeType
        
        if (encoding.starts(with: DATA_SCHEME)) {
            let clobWithoutScheme = encoding.replacingOccurrences(of: DATA_SCHEME, with: "")
            // split to place file mime type into first element
            let clobSplitOnComma = clobWithoutScheme.split(separator: ",")
            // this should contain ';base64' at end, so strip that as well
            self.dataMediaType = String(clobSplitOnComma[0].split(separator: ";")[0])
            self.data = String(clobSplitOnComma[1])
        }
        else {
            self.dataMediaType = ""
            self.data = encoding
        }
    }
    
    public func decode() throws -> Data {
        if let data = Data(base64Encoded: String(describing: self.data), options: .ignoreUnknownCharacters) {
            return data
        }
        
        throw Base64FileError.failDecoding
    }
}
